doc/devdocs/locks.rst

```diff
@@ -0,0 +1,127 @@
+.. currentmodule:: Base
+
+****************************************************
+Proper maintenance and care of multi-threading locks
+****************************************************
+
+The following strategies are used to ensure that the code is dead-lock free:
+
+ 1. structure code such that only one lock will need to be acquired at a time
+
+ 2. always acquire shared locks in the same order, as given by the table below
+
+ 3. avoid constructs that expect to need unrestricted recursion
+
+Below are all of the locks that exist in the system and the mechanism in use
+that prevents them from being a potential deadlock.
+
+The following are definitely leaf locks (level 1), and must not try to acquire any other lock:
+
+   * safepoint
+
+       Note that this lock is acquired implicitly by ``JL_LOCK`` and ``JL_UNLOCK``.
+       use the ``_NOGC`` variants to avoid that for level 1 locks.
+
+       While holding this lock, the code must not do any allocation or hit any safepoints.
+       Note that there are safepoint when doing allocation, enabling / disabling GC,
+       entering / restoring exception frame, and taking / releasing locks.
+
+   * shared_map
+   * finalizers
+   * pagealloc
+   * flisp
+
+       flisp itself is already threadsafe, this lock only protects the ``jl_ast_context_list_t`` pool
+
+
+The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
+
+   * symbol_table
+
+The following are level 3 locks, which can only acquire level 1 or level 2 locks internally:
+
+   * dump
+   * typecache
+
+The following is a level 4 locks, which can only recurse to acquire level 1, 2, or 3 locks:
+
+   * Method->writelock
+
+     but note that this is violated by staged functions!
+
+The following are level 5 locks, which can only recurse to acquire to acquire locks at lower levels:
+
+   * MethodTable->writelock
+
+     but note that this is violated by staged functions!
+
+The following is a proposed level 6 locks, which can only recurse to acquire locks at lower levels:
+
+   * staged
+
+       this theoretical lock would create a priority inversion from the `method->writelock` (level 4),
+       but only prohibiting only running any staging function in parallel
```

Comment:
what is staging then? is this related to staged functions or not? either way, what is the difference between "staged function" as used elsewhere vs "staging function" here?